src/fty_common_messagebus_unix_socket_helpers.cc

```diff
@@ -0,0 +1,185 @@
+
+        //add meta data
+        payload.push_back(BEGIN_METADATA);
+        for (const auto& pair : msg.metaData())
+        {
+            payload.push_back(pair.first);
+            payload.push_back(pair.second);
+        }
+        payload.push_back(END_METADATA);
+
+        //add user data
+        for(const auto& item : msg.userData())
+        {
+            payload.push_back(item);
+        }
+
+        sendFrames(socket, payload);
+    }
+
+    std::vector<std::string> recvFrames(int socket)
+    {
+        //format => [ Number of frames ], [ <size of frame 1> <data> ], ... [ <size of frame N> <data> ]
+
+        //get the number of frames
+        uint32_t numberOfFrame = 0;
+
+        
+        if(read(socket, &numberOfFrame, sizeof(uint32_t)) != sizeof(uint32_t))
+        {
+            throw MessageBusException("Error while reading number of frame");
+        }
+        
+
+        //Get frames
+        std::vector<std::string> frames;
+
+        for( uint32_t index = 0; index < numberOfFrame; index++)
+        {
+            //get the size of the frame
+            uint32_t frameSize = 0;
+            
+            
+            if(read(socket, &frameSize, sizeof(uint32_t)) != sizeof(uint32_t))
+            {
+                throw MessageBusException("Error while reading size of frame");
+            }
+            
+
+            if(frameSize == 0)
+            {
+                throw MessageBusException("Read error: Empty frame");
+            }
+
+            //get the payload of the frame
+            std::vector<char> buffer;
+            buffer.reserve(frameSize+1);
+            buffer[frameSize] = 0;
+
+            if(read(socket, &buffer[0], frameSize) != frameSize)
+            {
+                throw MessageBusException("Read error while getting payload of frame");
+            }
+            
+            std::string str(&buffer[0]);
+            
+            frames.push_back(str);
+        }
+        
+        return frames;
+    }
+    
+    void sendFrames(int socket, const std::vector<std::string> & payload)
+    {
+        //Send number of frame
+        uint32_t numberOfFrame = payload.size();
+        
+        if ( write(socket, &numberOfFrame, sizeof(uint32_t)) != sizeof(uint32_t) )
+        {
+            throw MessageBusException("Error while writing number of frame on socket "+std::to_string(socket));
+        }
+        
+        for(const std::string & frame : payload)
+        {
+            uint32_t frameSize = frame.length() + 1;
+            
+            if ( write(socket, &frameSize, sizeof(uint32_t)) != sizeof(uint32_t) )
+            {
+                throw MessageBusException("Error while writing size of frame on socket "+std::to_string(socket));
+            }
+            
+            if ( write(socket, frame.data(), frameSize) != frameSize )
+            {
+                throw MessageBusException("Error while writing payload on socket "+std::to_string(socket));
+            }    
```

Comment:
Instead of doing a bunch of `write()`s and having to deal with a bunch of potential failure points, compose the message into a buffer and perform a single `write()` at the end.